single_kernel_mongo/managers/config.py

```diff
@@ -0,0 +1,329 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Mongo configuration."""
+
+import logging
+import time
+from abc import ABC, abstractmethod
+from itertools import chain
+
+from ops import Container
+from typing_extensions import override
+
+from single_kernel_mongo.config.audit_config import AuditLog
+from single_kernel_mongo.config.literals import LOCALHOST, CharmRole, MongoPorts, Substrates
+from single_kernel_mongo.config.logrotate_config import LogRotateConfig
+from single_kernel_mongo.core.structured_config import MongoConfigModel, MongoDBRoles
+from single_kernel_mongo.core.workload import WorkloadBase
+from single_kernel_mongo.exceptions import WorkloadServiceError
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongodb_users import BackupUser, MonitorUser
+from single_kernel_mongo.workload import (
+    get_logrotate_workload_for_substrate,
+    get_mongodb_exporter_workload_for_substrate,
+    get_pbm_workload_for_substrate,
+)
+from single_kernel_mongo.workload.log_rotate_workload import LogRotateWorkload
+
+logger = logging.getLogger(__name__)
+
+
+class CommonConfigManager(ABC):
+    """A generic config manager for a workload."""
+
+    config: MongoConfigModel
+    workload: WorkloadBase
+    state: CharmState
+
+    def set_environment(self):
+        """Write all parameters in the environment variable."""
+        if self.workload.env_var != "":
+            parameters = chain.from_iterable(self.build_parameters())
+            self.workload.update_env(parameters)
+
+    def get_environment(self) -> str:
+        """Gets the environment for the defined service."""
+        env = self.workload.get_env()
+        return env[self.workload.env_var]
+
+    @abstractmethod
+    def build_parameters(self) -> list[list[str]]:  # pragma: nocover
```

Comment:
why is it needed to have a list of lists?

Reply:
There are builders methods that return only one parameter and some other that return a list of parameters.
Ensuring that all those builders returns a list allows to have a cleaner code which can the be chained in the final parameter string.
```python3
@override
def build_parameters(self) -> list[list[str]]:
    return [
        self.binding_ips,
        self.port_parameter,
        self.auth_parameter,
        self.tls_parameters,
        self.log_options,
        self.audit_options,
    ]
```